Bowling/Bowling/Frames.cs

```diff
@@ -0,0 +1,286 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Text;
+
+namespace BowlingApp
+{
+    public class Frames
+    {
+        public const int frameNumberTotal = 10;   // total number of frames in a bowling game
+        public const int bonusFrame = 10;    // frame where bonus roll could occur
+        private List<Frames> framesCollection = new List<Frames>();  // the collection of Frames objects
+        private int marksLength = 0;  // provides the total length of bowling scores string
+        private int placeCount = 0;  // placeCount will be used as a pointer in our bowling score string
+        private string score = string.Empty;  // string for each scoring mark in the bowling score string
+
+        public int frameNumber { get; set; }  // the frame number for these rolls
+        public int firstRoll { get; set; }  // first roll score
+        public int secondRoll { get; set; }  // second roll score
+        public int bonusRoll { get; set; }  //  bonus roll score
+        public string frameOutcome { get; set; }  // is final frame outcome a strike, spare, or open?
+        public int frameScore { get; set; }  //  total score for this frame
+
+        public List<Frames> CreateFrames(string bowlingMarks)
+        {
+            // get total length of the string entered.  This will help in ensuring our pointer doesn't look past the end of the string
+            marksLength = bowlingMarks.Length;
+
+            // run procedures to create an object from the Frames class for each of the 10 frames and add it to our frames List collection
+            try
+            {
+                // create loop that builds the 10 Frames objects
+                for (int i = 1; i <= frameNumberTotal; i++)
+                {
+                    // since the 10th frame can have potential bonus rolls, only the first 9 frames are contained in the procedure below.
+                    if ((placeCount <= marksLength) && (i != 10))
+                    {
+                        createFrames1To9(i, bowlingMarks);
+                    }
+                    else  // this procedure will handle only marks retrieved from the 10th frame.
+                    {
+                        createFrame10(i, bowlingMarks);
+                    }
+                }
+            }
+            catch (Exception ex)
+            {
+                Console.WriteLine(String.Concat("There was a problem creating the Frames collection: ", ex.Message));
+            }
+
+            return framesCollection;
+        }
+
+        /** Frame 1-9 Creation **/
+        private void createFrames1To9(int frameNumber, string bowlingMarks)
```

Comment:
You and I know why frames 1-9 are special, but for someone who has never bowled this name does not explain why it does not create all the frames.

Further, when I see a method named `createFoo`, I expect it to create and return a `Foo`; but this method returns `void`. This brings me to a bigger point about this class in particular: we aim to create classes that either hold data (plain old data object) or manipulate data (have function that manipulate data). This follows a functional programming paradigm even though we are using OO languages; but it gives us the same advantages you get with functional languages: it doesn't matter how many times I call a method with the same input it always produces the same output. This makes the code easier to reason about, and also makes it easier to test since there is less setup involved for each test case. I know this is a *big* ask, but see if you can rewrite this into classes that either hold data or act on data

Reply:
I count this method at around 80ish lines of code; while there is no hard and fast rule about some number of lines being too long, there are two considerations that can hint at a method being too long. First, how many things is it doing? If it's more than one, that's too many things. Second, can you (or someone who doesn't already know the code) read the method and quickly articulate what it does? Remember that source code is for human consumption, our goal is to make it readable for the next person that has to read/modify it.

Reply:
I want to piggy back off of this a little bit.

What if in the future we get an additional requirement that says we need to create a bowling game with only 5 frames?  This code is pretty rigid it how flexible it may be for making future modifications.  Maybe having `1to9` as a suffix should be reconsidered.  "Naming is hard" is a phrase we say a lot, but we spend a lot of time thinking about names for things.